common/src/main/java/io/druid/collections/LoadBalancingPool.java

```
@@ -0,0 +1,129 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.collections;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Supplier;
+import com.metamx.common.logger.Logger;
+
+import java.io.IOException;
+import java.util.Queue;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Simple load balancing pool that always returns the least used resource
+ *
+ * @param <T>
+ */
+public class LoadBalancingPool<T> implements Supplier<ResourceHolder<T>>
+{
+  private static final Logger log = new Logger(LoadBalancingPool.class);
+
+  private final Supplier<T> generator;
+
+  private final Queue<CountingHolder> objects;
+
+  private final int maxCapacity;
+  private final AtomicInteger size = new AtomicInteger(0);
+
+  public LoadBalancingPool(int maxCapacity, Supplier<T> generator)
+  {
+    Preconditions.checkArgument(maxCapacity > 0, "capacity must be greater than 0");
+    Preconditions.checkNotNull(generator);
+
+    this.generator = generator;
+    this.maxCapacity = maxCapacity;
+    this.objects = new PriorityBlockingQueue<>(maxCapacity);
+  }
+
+  public ResourceHolder<T> get()
+  {
+    CountingHolder obj;
+    if(size.getAndIncrement() > maxCapacity) {
+      // at capacity just take things from the queue
+      size.decrementAndGet();
+
+      // things never stay out of the queue for long, so we'll get it eventually
+      do { obj = objects.remove(); } while(obj == null);
```

I think ```objects.remove()``` is going to throw exceptions rather than return null when the pool is empty